MAINT: default to C11 rather than C99, fix most build warnings with Clang 14

rgommers opened the pull request:
The large number of build warnings on macOS was getting annoying, so I dug in and solved all but a handful of them (it was 51, now there are 5 left). Those last ones were tricky to address, and silencing `-Wunused-function` didn't seem quite right.

One main cause of warnings was the use of C11 features. Rather than trying to change that code, it made sense to start requiring C11 explicitly. The biggest change in C11 was to make previously-mandatory C99 features like `complex.h` and complex types optional - and we were already allowing that of course, because MSVC doesn't support complex types. Since all compilers we care about support C11, and the `typedef` redefinitions that we use in multiple places are a C11-only feature, documenting that C11 is what we require see and compiling with `-std=c11` seems fine.

This PR also cleans up uses of `#pragma GCC diagnostic`. Those should be avoided, especially for `-Wmaybe-uninitialized`, because Clang also looks at these pragmas and emits warnings when it doesn't understand them (and `-Wmaybe-uninitialized` doesn't exist there). There are usually better ways to deal with the problem, like changing the code that emits the warnings.

~Finally worth pointing out is one odd bug in `loops_minmax.dispatch.c.src` which disabled rather than enabled some code that was meant to run on Arm only. This was clearly a typo; tests all seem to pass; I didn't investigate performance.~

rgommers committed:
MAINT: improve debugging code in `umath_linalg.cpp` so it doesn't emit warnings

Also a simple fix for a `-Wmaybe-uninitialized` warning in
`special_integer_comparisons.cpp` (GCC isn't smart enough to
understand the `default` part of the switch statement without
the added line).

rgommers committed:
MAINT: fix comparison in npysort/selection.cpp so it doesn't emit a warning

That warning was, with clang on macOS:
```
../numpy/_core/src/npysort/selection.cpp:529:15: warning: result of comparison of constant 1 with expression of type 'NPY_SELECTKIND' is always false [-Wtautological-constant-out-of-range-compare]
    if (which >= NPY_NSELECTS) {
```

rgommers committed:
MAINT: fix unused function warning in `multiarray/item_selection.c`

The warning was:
```
[2/3] Compiling C object numpy/_core/_multiarray_umath.cpython-39-darwin.so.p/src_multiarray_item_selection.c.o
../numpy/_core/src/multiarray/item_selection.c:2371:1: warning: unused function 'count_nonzero_bytes_384' [-Wunused-function]
count_nonzero_bytes_384(const npy_uint64 * w)
```

rgommers committed:
BLD: use C11 rather than C99 as the C standard version we target

In effect we were already using C11 for a long time, because that's
where `complex.h` and complex types became optional rather than
mandatory. Also, there were several places where we used typedef
redefinitions, which is a C11 feature, as indicated by build warnings
like these:

```
In file included from ../numpy/_core/src/multiarray/common.h:6:
../numpy/_core/include/numpy/ndarraytypes.h:308:31: warning: redefinition of typedef 'NpyAuxData' is a C11 feature [-Wtypedef-redefinition]
typedef struct NpyAuxData_tag NpyAuxData;
                              ^
../numpy/_core/src/umath/loops.h.src:37:31: note: previous definition is here
typedef struct NpyAuxData_tag NpyAuxData;
                              ^

In file included from ../numpy/_core/src/multiarray/array_method.h:14:
../numpy/_core/include/numpy/_dtype_api.h:108:3: warning: redefinition of typedef 'PyArrayMethod_Context' is a C11 feature [-Wtypedef-redefinition]
} PyArrayMethod_Context;
  ^
../numpy/_core/src/umath/loops.h.src:36:42: note: previous definition is here
typedef struct PyArrayMethod_Context_tag PyArrayMethod_Context;

```

rgommers committed:
MAINT: remove GCC pragmas that cause warnings with Clang

The warning with Clang 14 is:

```
[3/4] Compiling C object numpy/_core/_multiarray_umath.cpython-39-darwin.so.p/meson-generated_einsum.c.o
../numpy/_core/src/multiarray/einsum.c.src:408:32: warning: unknown warning group '-Wmaybe-uninitialized', ignored [-Wunknown-warning-option]
                               ^

[173/301] Compiling C object numpy/_core/_multiarray_umath.cpython-39-darwin.so.p/meson-generated_scalarmath.c.o
../numpy/_core/src/umath/scalarmath.c.src:1810:36: warning: unknown warning group '-Wmaybe-uninitialized', ignored [-Wunknown-warning-option]
    #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
                                   ^
```

These pragmas are outdated and shouldn't be used. Instead, use
compiler flags (if really needed) that are feature-tested for actually
existing for the compiler that is used in the build.

rgommers committed:
MAINT: delete duplicate `simd_cabsolute_f32/f64` functions

These were yielding unused function warnings with Clang 14, and
turned out to indeed be unused. The code was an exact copy of
the same function in `loops_unary_complex_dispatch.c.src`, where
it was used.

seberg commented on numpy/_core/src/umath/loops_arithm_fp.dispatch.c.src:
```diff
@@ -272,61 +272,6 @@ simd_csquare_f64(npyv_f64 x)
-    return npyv_mul_@sfx@(hypot, larger);
-}
-#endif // VECTOR
-/**end repeat**/
```
For those not reading the commit message carefully:  A copy of this is in `loops_unary_complex.dispatch.c.src` where it is actually used.

seberg reviewed (dismissed):
LGTM, but that arm typo does *not* look like a typo, I can see it leading to ~uninitialized~ unused variables though.

I assume nobody will care about C11 vs. C99.  If anyone does (and actually has problems with those redefinitions), I suspect they can untangle the imports to fix it with some time.

rgommers commented on numpy/_core/src/umath/loops_minmax.dispatch.c.src:
```diff
@@ -350,7 +350,7 @@ NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(@TYPE@_@kind@)
         }
     // unroll scalars faster than non-contiguous vector load/store on Arm
-    #if !defined(NPY_HAVE_NEON) && @is_fp@
+    #if defined(NPY_HAVE_NEON) && @is_fp@
```
Hmm yeah, I misread that. I'll just drop the commit for now and revisit this separately. The SIMD code may need some `-Wno-xxx` flags, but those are hard to add per target right now.

rgommers commented:
One CI failure too, with Clang-cl on Windows:
```
numpy/random/_common.cp311-win_amd64.pyd.p/numpy/random/_common.pyx.c(25481,43): error: operand of type '_Fcomplex' (aka 'struct _C_float_complex') where arithmetic or pointer type is required
      return x + y*(__pyx_t_float_complex)_Complex_I;
```
Probably due to `-std=c11`, will check.

seberg commented:
Right, I just rebased gh-25044 I suspect that might help.

lysnikolaou commented:
The `clang-cl` failure probably is a Cython bug. It's like related to this [line](https://github.com/cython/cython/blob/master/Cython/Utility/Complex.c#L7), where Cython tries to detect c11 and it assumes that complex support is okay for c11 (which it is, but Cython's logic later on is to just fall back to using a struct for all MSVC-like environments).

Should we submit a bug report there?

lysnikolaou commented:
The Cython bug has been fixed in cython/cython#5809. Should we try Windows CI out with cython master?

rgommers commented:
> The Cython bug has been fixed in [cython/cython#5809](https://github.com/cython/cython/pull/5809). Should we try Windows CI out with cython master?

I tried this on my fork, CI passes now with both Clang-cl and MSVC with the Cython `3.0.x` branch. Once `3.0.6` is out we can move this forward.

charris commented:
I put a backport label on this, seems pretty straight forward.

rgommers committed:
BLD: update minimum Cython version to 3.0.6

Needed to include the bug fix for using complex types with C11
(Cython PR 5809).

rgommers commented:
This is happy now that the Cython issue for C11 was fixed and released in 3.0.6

> I put a backport label on this, seems pretty straight forward.

I'd be fine either way with backporting the MAINT commits yes or no. I'd prefer not to backport the BLD ones though.

charris commented on pyproject.toml:
```diff
@@ -2,7 +2,7 @@
 requires = [
     "meson-python>=0.15.0",
-    "Cython>=3.0",
+    "Cython>=3.0.6",  # keep in sync with version check in meson.build
```
Should `build_requirements.txt` also be updated with this version?

rgommers replied:
Good catch, updated.

lysnikolaou replied:
And maybe also update `environment.yml`?

rgommers replied:
No need I think, we don't add lower bounds or pins there unless there's an actual problem to resolve. Which won't be the case here, it should always pick the most recent Cython version.

lysnikolaou replied:
Oh okay, I though since there's a pin for `>=3.0`, an update might be needed.

rgommers replied:
I read that more as "no longer supports 0.29.X" - we should be able to drop that `<3.0` at some point, once other relevant packages that one may want to install into the same env have transitioned.

seberg commented:
Thanks Ralf, looks all good now, let's get this in.

charris commented:
> I'd be fine either way with backporting the MAINT commits yes or no.

I'm going to skip this. There are enough differences between 1.26.x and main that it feels risky. I plan on maybe two more 1.26 releases, and if we can get through those I'll be happy.

ngoldbaum commented:
Thanks for this, the warnings were indeed annoying!
